corehq/apps/cloudcare/utils.py

```diff
@@ -6,10 +6,70 @@
 from six.moves.urllib.parse import quote
 
 from corehq import toggles
-from corehq.apps.app_manager.dbaccessors import get_apps_in_domain
+from corehq.apps.app_manager.dbaccessors import (
+    get_app_ids_in_domain,
+    get_apps_in_domain,
+    get_latest_build_doc,
+    get_latest_build_id,
+    get_latest_released_app_doc,
+    get_latest_released_build_id,
+)
 from corehq.util.quickcache import quickcache
 
 
+def can_user_access_web_app(app, user, domain):
+    """
+    :param app: app doc (not wrapped Application)
+    :param user: either a WebUser or CommCareUser
+    :param domain: domain name
+    """
+    # Backwards-compatibility - mobile users haven't historically required this permission
+    has_access_via_permission = user.is_commcare_user()
+    if user.is_web_user() or user.can_access_any_web_apps(domain):
+        has_access_via_permission = user.can_access_web_app(domain, app.get('copy_of', app.get('_id')))
+
+    has_access_via_group = False
+    if toggles.WEB_APPS_PERMISSIONS_VIA_GROUPS.enabled(domain):
```

Comment:
> That'll make has_access_via_group always true for web users in domains with this flag on

I assume this isn't a problem in isolation (and is actually correct), but that because of the `or` logic below (`has_access_via_permission or has_access_via_group`), this is incorrect.

Previously the apps that a web user didn't have access to due to permissions would have been filtered out before getting to this code, whereas now the function attempts to check both at the same time. So the only time the group permission would have applied previously was when an app was **not** filtered out in the permissions filter (i.e., has_access_via_permission is true) and then the user did not have access via groups, so then the app was filtered out in that step.

All that to say, I believe this is the same precedence related change you highlighted above (`has_access_via_permission and has_access_via_group`) right?

Reply:
yes, that's right

Reply:
Should be resolved in 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02